### ISSUE TITLE
Composition's poly expected degree mismatch

### ISSUE THREAD

**Issue by 4l0n50** [novifinancial/winterfell#98]
Whenever there is a trace column with the same repeated entry, there's a mismatch on the expected degree of the composition poly as long as it is computed as `constrain_degree*(trace_length -1) - divisor_degree` here https://github.com/novifinancial/winterfell/blob/46dce1adf06dcb0778b546054e938f0144fc1288/prover/src/constraints/evaluation_table.rs#L440 and here https://github.com/novifinancial/winterfell/blob/46dce1adf06dcb0778b546054e938f0144fc1288/air/src/air/transition/degree.rs#L103

Although seems to be only a problem for debugging, it is annoying because it is not the expected behaviour. I wonder if this problem might also happen when some trace columns are generated, for example, as the evaluations of the identity polynomial

**Comment by irakliyk** [comment]
Yeah - this is an unfortunate consequence of using pre-defined degrees vs. actual degrees. There are two ways around it:

1. If it is possible to insert random values somewhere in the column, the degree of the column would be guaranteed to be `trace_length - 1`. I've done something like this in [one of the examples](https://github.com/novifinancial/winterfell/blob/main/examples/src/merkle/prover.rs#L86).
2. Another approach is to set the number of trace exceptions to 2 and then insert random values in all columns of the last row. This would also guarantee that degrees of all columns would be `trace_length - 1`.

It may also be possible to update the library in the future to rely on actual column degrees (instead of forcing them to always be `trace_length - 1`. Originally, I decided against doing this because I assumed that checking the degree would be pretty expensive. However, I've never actually benchmarked this - and it is possible that the extra expense is negligible in the grand scheme of things.

**Comment by 4l0n50** [comment]
Thanks for your answer! Then maybe this is just about changing the error message. Right now it is not possible to accept columns with degree < trace_length-1 as part of the witness, and the message should indicate so.
